zk-token-sdk/src/encryption/auth_encryption.rs

```diff
@@ -92,6 +102,77 @@ impl AeKey {
     pub fn decrypt(&self, ct: &AeCiphertext) -> Option<u64> {
         AuthenticatedEncryption::decrypt(self, ct)
     }
+
+    /// Reads a JSON-encoded key from a `Reader` implementor.
+    pub fn read_json<R: Read>(reader: &mut R) -> Result<Self, Box<dyn error::Error>> {
+        let bytes: [u8; 16] = serde_json::from_reader(reader)?;
+        Ok(Self(bytes))
+    }
+
+    /// Writes to a `Write` implementer with JSON-encoding.
+    pub fn write_json<W: Write>(&self, writer: &mut W) -> Result<String, Box<dyn error::Error>> {
+        let bytes = self.0;
+        let json = serde_json::to_string(&bytes.to_vec())?;
+        writer.write_all(&json.clone().into_bytes())?;
+        Ok(json)
+    }
+
+    /// Derive a key from an entropy seed.
+    ///
+    /// The seed is hashed using SHA3-512 and the first 16 bytes of the digest is taken as key. The
+    /// hash SHA3-512 is used to be consistent with `ElGamalKeypair::from_seed`.
+    pub fn from_seed(seed: &[u8]) -> Result<Self, Box<dyn error::Error>> {
+        const MINIMUM_SEED_LEN: usize = 16;
+
+        if seed.len() < MINIMUM_SEED_LEN {
+            return Err("Seed is too short".into());
+        }
+
+        let mut hasher = Sha3_512::new();
+        hasher.update(seed);
+        let result = hasher.finalize();
+
+        Ok(Self(result[..16].try_into()?))
+    }
+
+    /// Derive a key from a seed phrase and passphrase.
+    pub fn from_seed_phrase_and_passphrase(
+        seed_phrase: &str,
+        passphrase: &str,
+    ) -> Result<Self, Box<dyn error::Error>> {
+        Self::from_seed(&generate_seed_from_seed_phrase_and_passphrase(
+            seed_phrase,
+            passphrase,
+        ))
+    }
+}
```

Comment:
Same comment as in the ElgamalKeypair commit regarding duplicate methods with the same names.